Updated hub_demo to include string descriptors

stumpja opened the pull request:
The hub demo will now print usb string descriptors:

1] Manufacturer
2] Product Description
3] Serial Number

Also removed a new line return to make the output cleaner.

Please review.

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +239,83 @@ byte getconfdescr( byte addr, byte conf )
+ }
+ return( rcode );
+}
+
```
Please fix the indentation in this functions. Btw the "Auto Format" tool in the Arduino IDE is actually quite good. Just run that and it should fix the issue :)

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +239,83 @@ byte getconfdescr( byte addr, byte conf )
+  Usb.Task();
+  if( Usb.getUsbTaskState() >= 0x80 ) {  // state configuring or higher
+    USB_DEVICE_DESCRIPTOR buf; 
+    rcode = Usb.getDevDescr( addr, 0, 0x12, ( uint8_t *)&buf );
```
Could you replace 0x12 with DEV_DESCR_LEN, so it will become clearer. There is one more here as well: https://github.com/stumpja/USB_Host_Shield_2.0/blob/master/examples/hub_demo/hub_demo.ino#L119. It would be nice if you could fix that as well while you at it.

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +239,83 @@ byte getconfdescr( byte addr, byte conf )
+{ 
+  byte rcode;
+  Usb.Task();
+  if( Usb.getUsbTaskState() >= 0x80 ) {  // state configuring or higher
```
0x80 can also be replaced with ```USB_STATE_CONFIGURING```.

Lauszus commented:
Please see my inline comments. Also if you want it would be nice if all the ```byte``` types could be replaced with ```uint8_t```, as the ```byte``` type is deprecated.

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +239,83 @@ byte getconfdescr( byte addr, byte conf )
+//  function to get single string description
+unsigned int getstrdescr( unsigned int addr, byte idx )
+{
+ byte buf[ 66 ];
```
What is the reason for choosing this exact size? We might also want to add a check to make sure that the returned length does not exceed the buffer.

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +239,83 @@ byte getconfdescr( byte addr, byte conf )
+   return( rcode );
+ }
+ for( i = 2; i < length; i+=2 ) {
+   Serial.print((char) buf[i]);
```
Also a comment here that the string is unicode UTF-16 encoded might be useful, so people can figure out why we skip every second character.

stumpja commented:
@Lauszus I have made some minor changes, there are still unaddressed changes that need to be made. I will be looking into them later today.

Lauszus commented:
@stumpja thank you! Sorry for being so nitpicky, but I thought this would be a good time to fix some of the other issues in the example as well :)

stumpja commented:
@Lauszuz I never got around to cleaning up the remaining issues last Friday. However, I have addressed all your concerns mentioned above at this point. If I have not fixed one properly, or if you find something else, please let me know. I appreciate your "nitpicky-ness", honestly, and if we are going to update an example, we should do it right, so no worries there.

Your comment about the buffer size, this code was lifted from @felis  at the [circuits@home](https://www.circuitsathome.com/mcu/arduino-usb-host-part-3-descriptors) website. I never questioned it but after a little researching it seems that the maximum length of a string descriptor is 255 bytes since the first byte is reserved for the bLength variable. My points of reference can be found here:

[StackOverflow](http://stackoverflow.com/questions/7193645/how-long-is-the-string-of-manufacturer-of-a-usb-device)
[GitHub](https://github.com/nonolith/node-usb/issues/65)

They seem to agree so I updated the buffer length to 256 to accommodate the bLength variable and the UTF-16LE encoded string descriptor. Once we get this code where we want it, I will created another pull request updating the USB_desc example. Thanks for all your help with this!

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +234,83 @@ byte getconfdescr( byte addr, byte conf )
+unsigned int getstrdescr( unsigned int addr, uint8_t idx )
+{
+  uint8_t buf[ 256 ];
+  unsigned int rcode;
```
Please replace with ```uint8_t```

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +234,83 @@ byte getconfdescr( byte addr, byte conf )
+  unsigned int rcode;
+  uint8_t length;
+  uint8_t i;
+  unsigned short langid;
```
Please replace with ```uint16_t```

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +234,83 @@ byte getconfdescr( byte addr, byte conf )
+    Serial.print("Error retrieving LangID table ");
+    return ( rcode );
+  }
+  langid = word(buf[3], buf[2]);
```
Please don't use the ```word``` function, as it is not standard and might cause troubles on non-Arduino boards.

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -198,13 +194,13 @@ void printhubdescr(uint8_t *descrptr, uint8_t addr)
+  uint8_t rcode;
+  uint8_t descr_length;
+  uint8_t descr_type;
   unsigned int total_length;
```
Can you replace this with ```uint16_t``` as well?

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -238,6 +234,83 @@ byte getconfdescr( byte addr, byte conf )
+}
+
+//  function to get single string description
+unsigned int getstrdescr( unsigned int addr, uint8_t idx )
```
This should return an ```uint8_t``` and addr is just an ```uint8_t``` as well.

Lauszus commented:
@stumpja thanks for you hard work! I have added some other nitpicks, then it should be ready to merge :)

stumpja commented:
@Lauszus I've been preoccupied for a bit, but I have finally made the corrections mentioned above. Please look it over again.

Also, as far as I can tell the print_hex function is the only place where int's. Should we go ahead and swap these to uint8_t's as well?

I do apologize for this delay. Once we get this finalized, I will quickly update the USB_desc example.

Thanks!

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -293,7 +293,7 @@ unsigned int getstrdescr( unsigned int addr, uint8_t idx )
     return ( rcode );
   }
-  langid = word(buf[3], buf[2]);
+  langid = makeWord(buf[3], buf[2]);
```
Sorry, but don't use the makeWord function either, as this might not be available on non-Arduino boards. Simply just do it yourself:
```C++
(buf[3] << 8) | buf[2];
```

Lauszus commented on examples/hub_demo/hub_demo.ino:
```diff
@@ -420,3 +419,9 @@ void printProgStr(const char* str)
+uint16_t makeWord( uint8_t h, uint8_t l )
+{
+  return (h << 8) | l ;
+}
```
Please remove this, as this will cause conflicts when using Arduino: https://github.com/arduino/Arduino/search?utf8=%E2%9C%93&q=makeWord.

stumpja commented:
@Lauszus,

I'll eventually figure out what I am doing. Sorry. Update has been made.

Lauszus commented:
I just merged your PR manually: https://github.com/felis/USB_Host_Shield_2.0/commit/ef7d2e7bf5dda3051f68325e57653e307d3a90d6. Thank you :)
